File: FedRaft-Core/src/main/java/org/bupt/cad/fedraft/beans/NodeInfo.java
package org.bupt.cad.fedraft.beans;


import org.apache.commons.lang3.StringUtils;

import java.util.Objects;

public final class NodeInfo {

    private String ip;
    private int port;
    private int trainerPort;

    final private static long E16 = 1 << 16;
    final private static long E8 = 1 << 8;

    public static String idToIp(long id) { //添加静态方法,避免新建对象
        String[] fields = new String[4];
        id >>= 32;
        for (int i = 0; i < 4; i++) {
            fields[3 - i] = String.valueOf(id % E8);
            id >>= 8;
        }
        return StringUtils.join(fields, ".");
    }

    public static int idToPort(long id) {
        id >>= 16;
        return (int) (id % E16);
    }

    public static int idToTrainerPort(long id) {
        return (int) (id % E16);
    }


    public NodeInfo(String ip, int port, int trainerPort) {
        setIp(ip).setPort(port).setTrainerPort(trainerPort);
    }

    public NodeInfo(long id) {
        String[] fields = new String[4];
        int trainerPort = (int) (id % E16);
        id >>= 16;
        int port = (int) (id % E16);
        id >>= 16;
        for (int i = 0; i < 4; i++) {
            fields[3 - i] = String.valueOf(id % E8);
            id >>= 8;
        }
        setIp(StringUtils.join(fields, ".")).setPort(port).setTrainerPort(trainerPort);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        NodeInfo nodeInfo = (NodeInfo) o;
        return getIp().equals(nodeInfo.getIp()) && getPort() == nodeInfo.getPort() && getTrainerPort() == nodeInfo.getTrainerPort();
    }

    @Override
    public int hashCode() {
        return Objects.hash(ip, port, trainerPort);
    }

    public long getNodeId() {
        // xxx.xxx.xxx.xxx
        String[] fields = getIp().split("\\.");
        long id = 0L;
        for (String field : fields) {
            id = (id << 8) + Long.parseLong(field);
        }
        id = (id << 32) + ((long) getPort() << 16) + getTrainerPort();
        return id;
    }


    public String getIp() {
        return ip;
    }

    private NodeInfo setIp(String ip) {
        this.ip = ip;
        return this;
    }

    public int getPort() {
        return port;
    }

    private NodeInfo setPort(int port) {
        this.port = port;
        return this;
    }

    @Override
    public String toString() {
        return ip + ':' + port +
                ':' + trainerPort;
    }

    public int getTrainerPort() {
        return trainerPort;
    }

    private NodeInfo setTrainerPort(int trainerPort) {
        this.trainerPort = trainerPort;
        return this;
    }
}

File: Trainer/utils/NodeInfo.py

from functools import reduce

class NodeInfo(object):
    
    E16 = 1 << 16
    E8 = 1 << 8
    
    def __init__(self, ip: str, manager_port: int, trainer_port: int) -> None:
        self.ip = ip
        self.manager_port = manager_port
        self.trainer_port = trainer_port
        
    def get_id(self) -> int:
        
        fields = self.ip.split(".")
        
        node_id = reduce(lambda x, y: (x << 8) + int(y), fields, 0)
        node_id = (node_id << 32) + (self.manager_port << 16) + self.trainer_port
        
        return node_id
        
        
    def __str__(self) -> str:
        return f"NodeInfo[{self.ip}:{self.manager_port}:{self.trainer_port}]"
    
    @classmethod
    def id_to_NodeInfo(cls, id: int):
        fields = [i for i in range(4)]
        trainer_port = id % NodeInfo.E16
        id >>= 16
        manager_port = id % NodeInfo.E16
        id >>= 16
        for i in range(4):
            fields[3 - i] = str(id % NodeInfo.E8)
            id >>= 8
            
        return NodeInfo(".".join(fields), manager_port, trainer_port)
        
        
    @classmethod
    def id_to_ip(cls, id: int) -> str:
        fields = [i for i in range(4)]
        id >>= 32
        for i in range(4):
            fields[3 - i] = str(id % NodeInfo.E8)
            id >>= 8
        return ".".join(fields)
        
    @classmethod
    def id_to_manager_port(cls, id: int) -> int:
        return (id >> 16) % NodeInfo.E16
    
    @classmethod
    def id_to_trainer_port(cls, id: int) -> int:
        return id % NodeInfo.E16
    